### PR TITLE
8323832: Load JVMCI with the platform class loader

### PR DESCRIPTION
**Pull request by dougxc**
This PR changes `jdk.internal.vm.ci` such that it is loaded by the platform class loader instead of the boot class loader. This allows Native Image to load a version of JVMCI different than the version on top of which Native Image is running. This capability is demonstrated and tested by `LoadAlternativeJVMCI.java`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323832](https://bugs.openjdk.org/browse/JDK-8323832): Load JVMCI with the platform class loader (**Enhancement** - P4) ⚠️ Issue is not open.

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17520/head:pull/17520` \
`$ git checkout pull/17520`

Update a local copy of the PR: \
`$ git checkout pull/17520` \
`$ git pull https://git.openjdk.org/jdk.git pull/17520/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17520`

View PR using the GUI difftool: \
`$ git pr show -t 17520`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17520.diff">https://git.openjdk.org/jdk/pull/17520.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17520#issuecomment-1906487227)

### CODE REVIEW

File: src/java.base/share/lib/security/default.policy [hunk 1]
```diff
@@ -163,6 +163,10 @@ grant codeBase "jrt:/jdk.internal.le" {
     permission java.security.AllPermission;
 };
 
+grant codeBase "jrt:/jdk.internal.vm.ci" {
```
Comment (dougxc) [inline src/java.base/share/lib/security/default.policy:166]: This is required as JVMCI is no longer loaded by the boot loader but should retain all permissions.

**Comment by xxDark** [comment]
Hello. I'm not a reviewer but I read through the conversation in JIRA and saw this comment:
 > [~pwoegerer] currently has a Native Image patch where he creates a URLClassLoader whose parent is jdk.internal.loader.ClassLoaders.BOOT_LOADER (retrieved via reflection and use of required --add-exports and --add-opens command line options). That is, he's using the non-delegating approach you mention.

There is zero reason to do this.
Passing `null` as parent class loader would suffice as boot loader just uses `findBootstrapClassOrNull` in `JavaLangAccess` either way.

**Comment by dougxc** [comment]
> Passing `null` as parent class loader would suffice as boot loader just uses `findBootstrapClassOrNull` in `JavaLangAccess` either way

Thanks! I've simplified the test accordingly: 1642276ea22a5d789e01a5ecb1059d8c5c8be284

File: test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        for (int i = 0; i < testClasses.length; i++) {
+            String e = testClasses[i];
+            if (new File(e).isDirectory()) {
+                e = e + File.separator;
+            }
+            cp[i] = new URI("file:" + e).toURL();
+        }
+
+        ClassLoader pcl = ClassLoader.getPlatformClassLoader();
+        URLClassLoader ucl = new URLClassLoader(cp, null);
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java:54]: I am missing something here, a `URLClassLoader` first delegates to its parent before searching its URLs, so how does this not find the platform loader versions of the JVMCI classes?

**Review by dholmes-ora — COMMENTED** [review]
The actual changes to load JVMCI via the platform loader seem fine.

I'm still puzzled by the need to do this as any non-delegating classloader would have allowed this even if JVMCI were loaded by the bootloader. And I don't understand how your test is working as it is using a delegating classloader.

File: test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java (hunk 2 above)
Comment (dougxc) [reply test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java:54]: With `new URLClassLoader(cp, null)`, the URL loader delegates directly to the boot loader, by-passing the platform loader.

**Comment by dougxc** [comment]
> I'm still puzzled by the need to do this as any non-delegating classloader would have allowed this even if JVMCI were loaded by the bootloader.

As far as I understand, even a non-delegating classloader cannot redefine a class loaded by the boot loader. I modified the test to show this and get:
```
java.lang.LinkageError: loader LoadAlternativeJVMCI$1 @4a1f4d08 attempted duplicate class definition for jdk.vm.ci.meta.ResolvedJavaType. (jdk.vm.ci.meta.ResolvedJavaType is in unnamed module of loader LoadAlternativeJVMCI$1 @4a1f4d08, parent loader 'bootstrap')
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1023)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
	at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:524)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:427)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:421)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:714)
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:420)
	at LoadAlternativeJVMCI$1.loadClass(LoadAlternativeJVMCI.java:61)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
	at LoadAlternativeJVMCI.main(LoadAlternativeJVMCI.java:77)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
	at java.base/java.lang.Thread.run(Thread.java:1575)
```

Test modification:
```
diff --git a/test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java b/test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java
index dd63867e7c2..28a6fedca38 100644
--- a/test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java
+++ b/test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java
@@ -51,7 +51,14 @@ public static void main(String[] args) throws Exception {
         }

         ClassLoader pcl = ClassLoader.getPlatformClassLoader();
-        URLClassLoader ucl = new URLClassLoader(cp, null);
+        URLClassLoader ucl = new URLClassLoader(cp, null) {
+            protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+                if (!name.startsWith("jdk.vm.ci")) {
+                    return super.loadClass(name, resolve);
+                }
+                return findClass(name);
+            }
+        };

         String[] names = {
             "jdk.vm.ci.meta.ResolvedJavaType",
```

**Comment by olpaw** [comment]
> There is zero reason to do this. Passing `null` as parent class loader would suffice as boot loader just uses `findBootstrapClassOrNull` in `JavaLangAccess` either way.

Right, using `null` does the same thing. In the final version will use that instead of accessing private field `jdk.internal.loader.ClassLoaders#BOOT_LOADER`.

File: test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+public class LoadAlternativeJVMCI {
+    public static void main(String[] args) throws Exception {
+        String[] testClasses = System.getProperty("test.classes").split(File.pathSeparator);
+        URL[] cp = new URL[testClasses.length];
+        for (int i = 0; i < testClasses.length; i++) {
+            String e = testClasses[i];
+            if (new File(e).isDirectory()) {
+                e = e + File.separator;
+            }
+            cp[i] = new URI("file:" + e).toURL();
```
Comment (AlanBateman) [inline test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java:50]: This should be `cp[I] = file.toURI().toURL()` as a file path needs encoding to be URI path component.

File: test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java (hunk 2 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java:54]: \<face-palm>

Thanks Doug.

**Comment by dholmes-ora** [comment]
> As far as I understand, even a non-delegating classloader cannot redefine a class loaded by the boot loader. I modified the test to show this and get:
> 
> ```
> java.lang.LinkageError: loader LoadAlternativeJVMCI$1 @4a1f4d08 attempted duplicate class definition for jdk.vm.ci.meta.ResolvedJavaType. (jdk.vm.ci.meta.ResolvedJavaType is in unnamed module of loader LoadAlternativeJVMCI$1 @4a1f4d08, parent loader 'bootstrap')
> 	at java.base/java.lang.ClassLoader.defineClass1(Native Method)

Interesting. I'm not sure why that should be happening in this case. I can imagine a potential split-package issue with the bootloader that doesn't happen with the platform loader. I will look into it.

**Review by dholmes-ora — APPROVED** [review]

**Comment by xxDark** [comment]
> > I'm still puzzled by the need to do this as any non-delegating classloader would have allowed this even if JVMCI were loaded by the bootloader.
> 
> As far as I understand, even a non-delegating classloader cannot redefine a class loaded by the boot loader. I modified the test to show this and get:
> 
> ```
> java.lang.LinkageError: loader LoadAlternativeJVMCI$1 @4a1f4d08 attempted duplicate class definition for jdk.vm.ci.meta.ResolvedJavaType. (jdk.vm.ci.meta.ResolvedJavaType is in unnamed module of loader LoadAlternativeJVMCI$1 @4a1f4d08, parent loader 'bootstrap')
> 	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
> 	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1023)
> 	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
> 	at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:524)
> 	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:427)
> 	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:421)
> 	at java.base/java.security.AccessController.doPrivileged(AccessController.java:714)
> 	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:420)
> 	at LoadAlternativeJVMCI$1.loadClass(LoadAlternativeJVMCI.java:61)
> 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
> 	at LoadAlternativeJVMCI.main(LoadAlternativeJVMCI.java:77)
> 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
> 	at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
> 	at java.base/java.lang.Thread.run(Thread.java:1575)
> ```
> 
> Test modification:
> 
> ```
> diff --git a/test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java b/test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java
> index dd63867e7c2..28a6fedca38 100644
> --- a/test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java
> +++ b/test/hotspot/jtreg/compiler/jvmci/LoadAlternativeJVMCI.java
> @@ -51,7 +51,14 @@ public static void main(String[] args) throws Exception {
>          }
> 
>          ClassLoader pcl = ClassLoader.getPlatformClassLoader();
> -        URLClassLoader ucl = new URLClassLoader(cp, null);
> +        URLClassLoader ucl = new URLClassLoader(cp, null) {
> +            protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
> +                if (!name.startsWith("jdk.vm.ci")) {
> +                    return super.loadClass(name, resolve);
> +                }
> +                return findClass(name);
> +            }
> +        };
> 
>          String[] names = {
>              "jdk.vm.ci.meta.ResolvedJavaType",
> ```

It can. You need to check if class is already loaded by trying `findLoadedClass` first.

**Comment by AlanBateman** [comment]
> You're right. I had forgotten the intricacies of class loader delegation. The only hard constraint on loading a class in multiple loaders is that `java.*` classes [must (only) be loaded by the boot loader](https://github.com/openjdk/jdk/blob/bccd823c8e40863bed70ff5b24772843203871a5/src/java.base/share/classes/java/lang/ClassLoader.java#L904).

Just to add that this restriction was relaxed in Java 9 to allow java.* classes be defined by the platform class loader. The code that is linked to here throws if the class loader is not the platform class loader. There isn't a user accessible ClassLoader object for the boot loader and testing `this == null` doesn't make sense.

**Review by magicus — APPROVED** [review]
Build changes are trivially fine

**Comment by dougxc** [comment]
I'm closing this PR as the Native Image use case https://bugs.openjdk.org/browse/JDK-8323832 was opened for can be solved with an appropriately crafted custom loader that does not delegate loading of JVMCI classes.

Thanks for the reviews anyway, especially @xxDark for highlighting that this change is unnecessary.

**Comment by dholmes-ora** [comment]
> You need to check if class is already loaded by trying findLoadedClass first.

Thanks @xxDark . I knew it should work. :)

**Pull request closed without merging**